lib/spack/spack/prompt.py

```diff
@@ -0,0 +1,73 @@
+# Copyright 2013-2024 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+import os
+import textwrap
+
+from llnl.util.tty.color import colorize
+
+from spack.util.environment import EnvironmentModifications
+
+
+def prompt_modifications(prompt, shell, env=os.environ):
+    mods = EnvironmentModifications()
+
+    if shell == "fish" or shell == "pwsh" or shell == "bat":
+        # requires a function and can't be set with os.environ
+        pass
+    elif shell == "csh":
+        mods.set("SPACK_OLD_PROMPT", env.get("prompt", None))
+        mods.set("prompt", prompt)
+    else:
+        mods.set("SPACK_OLD_PS1", env.get("PS1", "$$$$"))
+        if "TERM" in env and "color" in env["TERM"]:
+            if "BASH" in env:
+                bash_color_prompt = colorize(f"@G{{{prompt}}}", color=True, enclose=True)
+                mods.set("PS1", f"{bash_color_prompt} {env.get('PS1','$ ')}")
+            else:
+                zsh_color_prompt = colorize(f"@G{{{prompt}}}", color=True, enclose=False, zsh=True)
+                mods.set("PS1", f"{zsh_color_prompt} {env.get('PS1', '$ ')}")
+        else:
+            mods.set("PS1", f"{prompt} {env.get('PS1', '$ ')}")
+
+    return mods
+
+
+def custom_prompt(prompt, shell):
+    cmds = ""
+    if shell == "csh":
+        cmds += "if (! $?SPACK_OLD_PROMPT ) "
+        cmds += 'setenv SPACK_OLD_PROMPT "${prompt}";\n'
+        cmds += 'set prompt="%s ${prompt}";\n' % prompt
+    elif shell == "fish":
+        if "color" in os.getenv("TERM", ""):
+            prompt = colorize(f"@G{prompt} " % prompt, color=True)
+    elif shell == "bat" or shell == "pwsh":
+        # TODO
+        pass
+    else:
+        bash_color_prompt = colorize(f"@G{{{prompt}}}", color=True, enclose=True)
+        zsh_color_prompt = colorize(f"@G{{{prompt}}}", color=True, enclose=False, zsh=True)
+        cmds += textwrap.dedent(
+            rf"""
+            if [ -z ${{SPACK_OLD_PS1+x}} ]; then
+                if [ -z ${{PS1+x}} ]; then
+                    PS1='$$$$';
+                fi;
+                export SPACK_OLD_PS1="${{PS1}}";
+            fi;
+            if [ -n "${{TERM:-}}" ] && [ "${{TERM#*color}}" != "${{TERM}}" ] && \
+               [ -n "${{BASH:-}}" ];
+            then
+                export PS1="{bash_color_prompt} ${{PS1}}";
+            elif [ -n "${{TERM:-}}" ] && [ "${{TERM#*color}}" != "${{TERM}}" ] && \
+                 [ -n "${{ZSH_NAME:-}}" ];
+            then
+                export PS1="{zsh_color_prompt} ${{PS1}}";
+            else
+                export PS1="{prompt} ${{PS1}}";
+            fi
+            """
+        ).lstrip("\n")
```

Comment:
Is there any reason why we do this logic as shell functions rather than just inspecting the shell like I do in the function above? Seems like all of this could just get deleted and go into the environment mod. The `fish` prompt change doesn't even do anything at the moment...

Reply:
Two things:

1. You're not guaranteed that the environment in `spack` is the same environment as that of the user's shell, or even really that `spack` was launched from a shell. This might be pathological, but you could imagine a csh user or a sysadmin wrapping spack in bash scripts that are called from some *other* shell. If you inspect spack's environment in that situation, you get the wrong answer.
2. This logic has to save state in the user's shell environment so that we can put things back the way they were after deactivation.  And the saved state (the prompt) needs to come from the user's shell, not `spack`'s environment.

Reply:
Here's an illustrative example -- the shell detection stuff is generally not available from the environment of things launched from that shell:

```console
(spackbook):~$ echo $BASH
/bin/bash
(spackbook):~$ python -c 'import os; os.environ["BASH"]'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "<frozen os>", line 679, in __getitem__
KeyError: 'BASH'
```

```console
(spackbook):~> echo $ZSH_NAME
zsh
(spackbook):~> python -c 'import os; os.environ["ZSH_NAME"]'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "<frozen os>", line 679, in __getitem__
KeyError: 'ZSH_NAME'
```

Reply:
Okay yeah I noticed this in that the current prompt is not getting preserved with the dive. So how do we decide what goes into an `EnvironmentModification` object vs what goes into the shell wrapper?  Trial and error?

Reply:
The save state (bullet 2) seems like it would apply either way since we are dumping all of the env mods to a shell wrapper anyways...

Reply:
> So how do we decide what goes into an EnvironmentModification object vs what goes into the shell wrapper? Trial and error?

if you need to read directly from some variable in the user's environment that's *only* in the shell, you have to code things up like this.  This is not most things (e.g. PATH, etc. are preserved).  There are some nasty interactions to be aware of, e.g. macOS clears out some variables (DYLD_LIBRARY_PATH) on process launch -- see setup-env.sh for more on that) and Spack sanitizes the build environment--not sure that is relevant here but see #46823.